### DISCUSSION TITLE
On your work arXiv:2206.07044

### DISCUSSION THREAD

**Discussion by refraction-ray** [jcmgray/cotengra#14] [General]
_Hyper-optimized compressed contraction of tensor networks with arbitrary geometry_: great work and it is very impressive and helpful!

I am just curious whether cotengra now supports some of the features described in this work, since I find some ``compressed`` words in commit history and the file name. Besides, do you have any plan to open source the implementation of this paper as a great tool (either integrated with cotengra or as an independent package) in the future?

Also a technical side question, how do you compare this general approximation contraction scheme with MPS-TEBD scheme (eg. https://journals.aps.org/prx/abstract/10.1103/PhysRevX.10.041038) in the context of approximated quantum circuit simulation (say still the random circuit supremacy case) in terms of efficiency and approximation error.

Sorry if you find this post irrelevant from issues here, feel free to close this, and we can communicate via email if you prefer.

**Comment by jcmgray** [comment]
*(I've enabled dicussions and moved this as seems appropriate, hope thats ok)*

Thanks for the kind words about the paper. Regarding the first point:

1. Yes the tree generators are in `cotengra` already (use via `ctg.HyperCompressedOptimizer`), just not documented.
2. The compressed contraction functionality requires `quimb` (`tn.contract_compressed`), also not documented yet

That being said, there can be quite a lot of arcane tricks to applying it successfully, and so I'm not intending to push it publicly until I've written some sort of guide that covers these and FAQ etc. But certainly the point is to help people to try and use it soon!

With regard to quantum circuits - applying this to quantum circuits is obviously an interesting Q that we will look at soon, but felt was really too big of a topic to include in the initial paper. The paper you mention, and others, I think give some hope that there is a degree of approximability even for random quantum circuits, even if its just a 'polynomial' speedup. A naive problem is that the error of approximate contraction is not currently estimatable a priori, and when you use things like looking for convergence in `chi`, you rely on the error being smaller than the value. Whereas the target fidelities for the random circuits are e.g. 0.002 == 99.8% error.
